File: 2018/src/se/johannalynn/adventofcode/Day1.java
package se.johannalynn.adventofcode;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.*;

public class Day1 {
    public static void main(String[] args) throws FileNotFoundException {
        System.out.println("Day1");

        String inFileName = "2018/input/day1.txt";
        Scanner in = new Scanner(new File(inFileName));

        //star1(in);
        star2(in);
    }

    private static void star1(Scanner in) {
        int start = 0;
        while(in.hasNextLine()) {
            int frequency = Integer.parseInt(in.nextLine());
            start += frequency;
        }
        System.out.println(start);
    }

    private static void star2(Scanner in) {
        List<Integer> shifts = new ArrayList<>();
        while(in.hasNextLine()) {
            int frequency = Integer.parseInt(in.nextLine());
            shifts.add(frequency);
        }

        int current = 0;
        Set<Integer> frequencies = new HashSet<>();
        frequencies.add(current);
        int size = frequencies.size();

        boolean found = false;
        int i = 0;
        while(!found) {
            int frequency = shifts.get(i);
            current += frequency;
            frequencies.add(current);

            int newSize = frequencies.size();
            if (size == newSize) {
                System.out.println(current);
                found = true;
            }
            size = newSize;
            i++;
            if (i >= shifts.size()) {
                i = 0;
            }
        }
    }
}

File: src/se/johannalynn/adventofcode/y2018/Day1.kt
package se.johannalynn.adventofcode.y2018

import java.io.File
import java.io.FileNotFoundException
import java.util.*

object Day1 {

    @JvmStatic
    fun main(args: Array<String>) {
        println("Day1")

        val inFileName = "input/2018/day1.txt"
        val scanner = Scanner(File(inFileName))

        //star1(scanner);
        star2(scanner)

        scanner.close()
    }

    private fun star1(scanner: Scanner) {
        var start = 0
        while (scanner.hasNextLine()) {
            val frequency = Integer.parseInt(scanner.nextLine())
            start += frequency
        }
        println(start)
    }

    private fun star2(scanner: Scanner) {
        val shifts = ArrayList<Int>()
        while (scanner.hasNextLine()) {
            val frequency = Integer.parseInt(scanner.nextLine())
            shifts.add(frequency)
        }

        var current = 0
        val frequencies = HashSet<Int>()
        frequencies.add(current)
        var size = frequencies.size

        var found = false
        var i = 0
        while (!found) {
            val frequency = shifts[i]
            current += frequency
            frequencies.add(current)

            val newSize = frequencies.size
            if (size == newSize) {
                println(current)
                found = true
            }
            size = newSize
            i++
            if (i >= shifts.size) {
                i = 0
            }
        }
    }
}
